MapboxNavigation/InstructionsCardContainerView.swift

```diff
@@ -0,0 +1,306 @@
+            nextBannerView.instructionLabel.normalFont = style.nextBannerInstructionLabelNormalFont
+            nextBannerView.instructionLabel.shieldHeight = style.nextBannerInstructionLabelNormalFont.pointSize
+        }
+        
+        if let lanesView = view as? LanesView, let stackView = lanesView.subviews.first as? UIStackView {
+            let laneViews: [LaneView] = stackView.subviews.compactMap { $0 as? LaneView }
+            laneViews.forEach { laneView in
+                guard laneView.isValid else { return }
+                laneView.primaryColor = self.style.lanesViewDefaultColor
+                laneView.secondaryColor = self.style.lanesViewDefaultColor
+            }
+        }
+        
+        view.layoutIfNeeded()
+        
+        return view
+    }
+    
+    override public func layoutSubviews() {
+        super.layoutSubviews()
+    }
+    
+    private func gradientLayer(for view: UIView, with colors:[CGColor]? = nil) -> CAGradientLayer? {
+        guard !view.isHidden, let sublayers = view.layer.sublayers,
+              let firstLayer = sublayers.first as? CAGradientLayer,
+              let layerColors = firstLayer.colors as? [CGColor], layerColors.count == 2 else {
+            return nil
+        }
+        
+        if let colors = colors {
+            let colorsMatched = layerColors.reduce(false) { $0 || colors.contains($1) }
+            return colorsMatched ? firstLayer : nil
+        }
+        
+        return firstLayer
+    }
+    
+    public func updateInstruction(for step: RouteStep, distance: CLLocationDistance, previewEnabled: Bool = false) {
+        instructionsCardView.updateDistanceFromCurrentLocation(distance)
+        instructionsCardView.step = step
+        
+        // TODO: Merge Instructions Card, Lanes & Next Banner View Instructions
+        guard let instruction = step.instructionsDisplayedAlongStep?.last else { return }
+        updateInstruction(instruction, previewEnabled: previewEnabled)
+    }
+    
+    public func updateInstruction(_ instruction: VisualInstructionBanner, previewEnabled: Bool = false) {
+        if lanesView.isHidden {
+            lanesView.update(for: instruction)
+        } else if lanesView.isCurrentlyVisible && instruction.tertiaryInstruction == nil {
+            lanesView.hide()
+        }
+        
+        if nextBannerView.isHidden {
+            nextBannerView.update(for: instruction)
+            nextBannerView.instructionDelegate = self
+        } else if nextBannerView.isCurrentlyVisible && instruction.tertiaryInstruction?.text == nil {
+            nextBannerView.hide()
+        }
+        
+        if !previewEnabled, let distance = instructionsCardView.distanceFromCurrentLocation {
+            let highlightEnabled = distance < InstructionsCardConstants.highlightDistance
+            updateBackgroundColor(highlightEnabled: highlightEnabled)
```

Comment:
Every time the collection view is updated, we’re recalculating the distances for all remaining steps from the current location to the end of the leg, just to determine whether to highlight a card. This is suboptimal for a few reasons:

* It’s unlikely that the ten steps out is near enough to warrant highlighting, so we don’t need to calculate anything for it.
* Each of the distances has to be recalculated on every location update.

Taking a step back, the point of the highlighting feature is to highlight nearby maneuvers. There’s a simpler approach: on each location update, get the current step and mark it as nearby. If there’s less than _nearbyDistance_ meters left on the step, mark the next maneuver as nearby and subtract the distance to that maneuver from _nearbyDistance_, and so on.

Reply:
Yes, the current implementation isn't optimal. The suggested approach will certainly work if we assume 1 - 2 instruction cards are displayed at a time. 

Hence, we need to account for only visible cells for the current leg to ensure we persist the intended user experience.

Reply:
We don’t need to assume the number of cards visible. We’d start out by unhighlighting all the cards and iterating from the current step onward.